timesource/timesource.go

```diff
@@ -144,19 +148,57 @@ type NTPTimeSource struct {
 	timeQuery         ntpQuery // for ease of testing
 	now               func() time.Time
 
-	quit    chan struct{}
+	quit chan struct{}
+
+	stateMu sync.Mutex
 	started bool
 
-	mu           sync.RWMutex
-	latestOffset time.Duration
+	timeDataMu    sync.RWMutex
+	latestOffset  time.Duration
+	lastMonotonic time.Time
+	lastWallTime  time.Time
 }
 
 // Now returns time adjusted by latest known offset
+// and detects system time changes
 func (s *NTPTimeSource) Now() time.Time {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-	n := s.now()
-	return n.Add(s.latestOffset)
+	s.timeDataMu.RLock()
+
+	currentTime := s.now()
+	adjustedTime := currentTime.Add(s.latestOffset)
+
+	// Skip time change detection if time tracking not initialized yet
+	if s.lastMonotonic.IsZero() {
+		s.timeDataMu.RUnlock()
+		return adjustedTime
+	}
+
+	// Check for time inconsistency
+	monotonicElapsed := time.Since(s.lastMonotonic)
+	wallClockElapsed := time.Duration(currentTime.UnixNano() - s.lastWallTime.UnixNano())
+	timeDiff := monotonicElapsed - wallClockElapsed
+
+	s.timeDataMu.RUnlock()
+
+	// If significant time change detected, update offset synchronously
+	if timeDiff.Abs() > TimeChangeThreshold {
+		logutils.ZapLogger().Warn("System time change detected",
+			zap.Duration("difference", timeDiff),
+			zap.Duration("threshold", TimeChangeThreshold))
+
+		// Synchronously update the NTP offset
+		if err := s.updateOffset(); err != nil {
+			logutils.ZapLogger().Error("Failed to update time offset after time change", zap.Error(err))
```

Comment:
This log call seems to be redundant because `updateOffset` already logs on error.